### DISCUSSION TITLE
Allow a lambda expression with a statement body to be converted to an expression tree

### DISCUSSION THREAD

**Discussion by maurosampietro** [dotnet/csharplang#4569] [Language Ideas]
Error CS0834 says: A lambda expression with a statement body cannot be converted to an expression tree.

So this is possible:

```c#
Expression<Func<string, bool>> example1 = str => str == "ok";`
```

But this is not possible:

```c#
    Expression<Func<string, bool>> example2 = str => 
    { 
         if( str == "ok" ) 
             return true; 

         return false
    };
```

However creating a new method and passing in the params does the trick:

```c#
    Expression<Func<string, bool>> example3 = str => RelayMethod( str );

    private bool RelayMethod( string str )
    {
        if( str == "ok" )
            return true;

        return false;
    }

```
The trick always works in any scenario so why don't we automatically generate the 'method trick' (when a lambda expression with a statement body is detected) and remove this limit?

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/issues/2545

**Comment by jnm2** [comment]
> The trick always works in any scenario so why don't we automatically generate the 'method trick' (when a lambda expression with a statement body is detected) and remove this limit?

If you look at the expression tree, it will no longer contain the operations for `str == "ok"`, just an opaque call to the .NET method "RelayMethod." That would be highly problematic in most of the places where expression trees are used, like Entity Framework.

**Comment by joserubenh** [comment]
Expressions already supports this.
I think the reason is not allowed is only because it cannot be compiled in some cases.
But that's the whole point of expressions, allow the final implementator to deal with it.

In my case, I'd like to analyze the expression tree to build an FPGA real-time analizer. Even if it cannot be compiled in csharp, an Expression tree from Expression<Action> would be incredibly useful.
